features/user-profile.md

```diff
@@ -0,0 +1,226 @@
+private profile.
+
+The following table gives an example of the field-based ACL of user record
+fulfilling: 1) both username, email and status are readable and discoverable;
+2) profile photo is readable, but not queryable; 3) game score is queryable;
+and 4) birthday is private:
+
+| Class | UserRole |    Field    | AccessLevel | DiscoveryLevel |
+|-------|----------|-------------|-------------|----------------|
+| *     | Public   | *           | ReadWrite   | Queryable      |
+| User  | Owner    | *           | ReadWrite   | Queryable      |
+| User  | AnyUser  | username    | ReadOnly    | Discoverable   |
+| User  | AnyUser  | email       | ReadOnly    | Discoverable   |
+| User  | AnyUser  | status      | ReadOnly    | Discoverable   |
+| User  | AnyUser  | profile_pic | ReadOnly    | NotQueryable   |
+| User  | AnyUser  | game_score  | ReadOnly    | Queryable      |
+| User  | AnyUser  | birthday    | NoAccess    | NotQueryable   |
+
+## Authentication Object
+
+The new authentication object (i.e. `_auth` object) is to replace the old
+ `_user` object, to avoid ambiguity with the user record.
+
+The authentication object contains the following fields:
+
+- `id (text)`: should be the same as `user._id` as a convention
+- `password (text)`: the password for authentication
+- `provider_info (jsonb)`: the authentication information from Authentication
+  Provider
+- `token_valid_since`: the timestamp limit the validity of the authentication
+  token
+- `last_login_at`: the timestamp when the user last login
+- `last_seen_at`: the timestamp when user last perform an action
+
+## New User Discovery API
+
+After revamping the user record and authentication object, a new user discovery
+API is proposed as a helper method query user records.
+
+The API signature will be `container.discoverUser(discoverData)` while
+`discoverData` is an object where each key-value pair will be transformed to
+an equality query predicate. The following snippet shows how it is transformed:
+
+```js
+const skygear = require('skygear');
+
+// the new User Discovery API
+skygear.discoverUser({ gender: 'male' })
+.then((users) => {
+  console.log(`${users.length} users founded`);
+});
+
+// the above equals to:
+const User = skygear.Record.extend('user');
+const q = skygear.Query(User);
+q.equalTo('gender', 'male');
+skygear.publicDB.query(q)
+.then((users) => {
+  console.log(`${users.length} users founded`);
+});
+```
+
+Also, the following user discovery methods are proposed as they are commonly
+used:
+
+- `container.discoverUserByUsername(username)`
+- `container.discoverUserByUsernames(usernames)`
+- `container.discoverUserByEmail(email)`
+- `container.discoverUserByEmails(emails)`
+
+## New Sign Up and Log In API
+
+The new sign up API is proposed to enable: 1) sign up with custom data;
+2) update user profile during sign up. The new sign up API will perform:
+
+1. Discover user with the provided data
```

Comment:
If the column of the auth data does not exists, it will be regarded as "no user is found" on dev mode, regarded as request error on production mode.